Allow unsafe through inline const

nbdd0121 opened the pull request:
Handle similar to closures.

Address https://github.com/rust-lang/rust/pull/104087#issuecomment-1324173328

Note that this PR does not fix the issue for `unsafe { [0; function_requiring_unsafe()] }`. This is fundamentally unfixable for MIR unsafeck IMO.

This PR also does not fix unsafety checking for inline const in pattern position. It actually breaks it, allowing unsafe functions to be used in inline const in pattern position without unsafe blocks. Inline const in pattern position is not visible in MIR so ignored by MIR unsafety checking (currently it is also not checked by borrow checker, which is the reason why it's considered an incomplete feature).

@rustbot label: +T-lang +F-inline_const

nbdd0121 committed:
Allow unsafe through inline const

This is handled similar to closures

nbdd0121 committed:
Allow unsafe through inline const for THIR unsafety checker

The closure handling code is changed slightly to avoid allocation
when THIR building failed.

nbdd0121 committed:
Add tests

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3595857228/jobs/6055849453) [(plain)](https://github.com/rust-lang/rust/commit/c4139471ee532df23e32bb201a0cf912cb0ff1b5/checks/9831105066/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- [mir-opt] src/test/mir-opt/building/custom/consts.rs stdout ----

error: compilation failed!
status: exit status: 101
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/mir-opt/building/custom/consts.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "-O" "-Copt-level=1" "-Zdump-mir=all" "-Zvalidate-mir" "-Zdump-mir-exclude-pass-number" "-Zmir-pretty-relative-line-numbers=yes" "-Zmir-opt-level=4" "-Zdump-mir-dir=/checkout/obj/build/x86_64-unknown-linux-gnu/test/mir-opt/building/custom/consts" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/mir-opt/building/custom/consts" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/mir-opt/building/custom/consts/auxiliary"
stdout: none
--- stderr -------------------------------
error: internal compiler error: /checkout/compiler/rustc_middle/src/mir/mod.rs:615:39: unwrapping cross-crate data
thread 'rustc' panicked at 'Box<dyn Any>', /checkout/compiler/rustc_errors/src/lib.rs:1575:9
stack backtrace:
stack backtrace:
   0:     0x7ffaf1c87ed5 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h197489db1f8041c9
   1:     0x7ffaf1cf7ac8 - core::fmt::write::h0810bb2cf209cb58
   2:     0x7ffaf1c79c81 - std::io::Write::write_fmt::hdfa2531741826335
   3:     0x7ffaf1c87ce1 - std::sys_common::backtrace::print::h005d089afd8691f6
   4:     0x7ffaf1c8b044 - std::panicking::default_hook::{{closure}}::h27d4d6f273ecdfab
   5:     0x7ffaf1c8ad0a - std::panicking::default_hook::hd5fc0b1f5e5bd948
   6:     0x7ffaf26d04c4 - rustc_driver[5df8529dbce8ebc5]::DEFAULT_HOOK::{closure#0}::{closure#0}
   7:     0x7ffaf1c8b7b4 - std::panicking::rust_panic_with_hook::hd84df5858a1fbe34
   8:     0x7ffaf53d5c23 - std[4d8c08a1d6cd93f8]::panicking::begin_panic::<rustc_errors[7a3c0cce756dd281]::ExplicitBug>::{closure#0}
   9:     0x7ffaf53ce106 - std[4d8c08a1d6cd93f8]::sys_common::backtrace::__rust_end_short_backtrace::<std[4d8c08a1d6cd93f8]::panicking::begin_panic<rustc_errors[7a3c0cce756dd281]::ExplicitBug>::{closure#0}, !>
  10:     0x7ffaf2670bd6 - std[4d8c08a1d6cd93f8]::panicking::begin_panic::<rustc_errors[7a3c0cce756dd281]::ExplicitBug>
  11:     0x7ffaf53c8e16 - std[4d8c08a1d6cd93f8]::panic::panic_any::<rustc_errors[7a3c0cce756dd281]::ExplicitBug>
  12:     0x7ffaf53c6d30 - <rustc_errors[7a3c0cce756dd281]::HandlerInner>::bug::<&alloc[61e56d2e13f18c62]::string::String>
  13:     0x7ffaf53c68f0 - <rustc_errors[7a3c0cce756dd281]::Handler>::bug::<&alloc[61e56d2e13f18c62]::string::String>
  14:     0x7ffaf556b915 - rustc_middle[114b3b537d7176f2]::ty::context::tls::with_context_opt::<rustc_middle[114b3b537d7176f2]::ty::context::tls::with_opt<rustc_middle[114b3b537d7176f2]::util::bug::opt_span_bug_fmt<rustc_span[e0114d164763424]::span_encoding::Span>::{closure#0}, ()>::{closure#0}, ()>
  15:     0x7ffaf556d5d9 - rustc_middle[114b3b537d7176f2]::util::bug::opt_span_bug_fmt::<rustc_span[e0114d164763424]::span_encoding::Span>
  16:     0x7ffaf2679908 - rustc_middle[114b3b537d7176f2]::util::bug::bug_fmt
  17:     0x7ffaf32e5812 - <rustc_middle[114b3b537d7176f2]::mir::ClearCrossCrate<&rustc_middle[114b3b537d7176f2]::mir::SourceScopeLocalData>>::assert_crate_local
  18:     0x7ffaf331514d - <rustc_mir_transform[5c4de65d0c4fc280]::check_unsafety::UnsafetyChecker>::register_violations::<&alloc[61e56d2e13f18c62]::vec::Vec<rustc_middle[114b3b537d7176f2]::mir::query::UnsafetyViolation>, core[227267dacc15a5d0]::iter::adapters::copied::Copied<std[4d8c08a1d6cd93f8]::collections::hash::set::Iter<rustc_hir[a42e12c5fc8ec2f6]::hir_id::HirId>>>
  19:     0x7ffaf33143ff - <rustc_mir_transform[5c4de65d0c4fc280]::check_unsafety::UnsafetyChecker as rustc_middle[114b3b537d7176f2]::mir::visit::Visitor>::visit_operand
  20:     0x7ffaf3314097 - <rustc_mir_transform[5c4de65d0c4fc280]::check_unsafety::UnsafetyChecker as rustc_middle[114b3b537d7176f2]::mir::visit::Visitor>::visit_rvalue
  21:     0x7ffaf3315d91 - rustc_mir_transform[5c4de65d0c4fc280]::check_unsafety::unsafety_check_result
  22:     0x7ffaf3311b40 - <rustc_mir_transform[5c4de65d0c4fc280]::check_unsafety::provide::{closure#0} as core[227267dacc15a5d0]::ops::function::FnOnce<(rustc_middle[114b3b537d7176f2]::ty::context::TyCtxt, rustc_span[e0114d164763424]::def_id::LocalDefId)>>::call_once
  23:     0x7ffaf451b232 - rustc_query_system[2b8658e7d9d902a0]::query::plumbing::try_execute_query::<rustc_query_impl[d90372ddb6f5ed6d]::plumbing::QueryCtxt, rustc_query_system[2b8658e7d9d902a0]::query::caches::VecCache<rustc_span[e0114d164763424]::def_id::LocalDefId, &rustc_middle[114b3b537d7176f2]::mir::query::UnsafetyCheckResult>>
  24:     0x7ffaf45f8566 - rustc_query_system[2b8658e7d9d902a0]::query::plumbing::get_query::<rustc_query_impl[d90372ddb6f5ed6d]::queries::unsafety_check_result, rustc_query_impl[d90372ddb6f5ed6d]::plumbing::QueryCtxt>
  25:     0x7ffaf4195b40 - <rustc_query_impl[d90372ddb6f5ed6d]::Queries as rustc_middle[114b3b537d7176f2]::ty::query::QueryEngine>::unsafety_check_result
  26:     0x7ffaf329ec3e - rustc_mir_transform[5c4de65d0c4fc280]::mir_const
  27:     0x7ffaf449815a - rustc_query_system[2b8658e7d9d902a0]::query::plumbing::try_execute_query::<rustc_query_impl[d90372ddb6f5ed6d]::plumbing::QueryCtxt, rustc_query_system[2b8658e7d9d902a0]::query::caches::DefaultCache<rustc_middle[114b3b537d7176f2]::ty::WithOptConstParam<rustc_span[e0114d164763424]::def_id::LocalDefId>, &rustc_data_structures[89239e8ccb8fd3af]::steal::Steal<rustc_middle[114b3b537d7176f2]::mir::Body>>>
  28:     0x7ffaf4636674 - rustc_query_system[2b8658e7d9d902a0]::query::plumbing::get_query::<rustc_query_impl[d90372ddb6f5ed6d]::queries::mir_const, rustc_query_impl[d90372ddb6f5ed6d]::plumbing::QueryCtxt>
  29:     0x7ffaf4181154 - <rustc_query_impl[d90372ddb6f5ed6d]::Queries as rustc_middle[114b3b537d7176f2]::ty::query::QueryEngine>::mir_const
  30:     0x7ffaf329fb8d - rustc_mir_transform[5c4de65d0c4fc280]::mir_promoted
  31:     0x7ffaf45cb2c4 - rustc_query_system[2b8658e7d9d902a0]::query::plumbing::get_query::<rustc_query_impl[d90372ddb6f5ed6d]::queries::mir_promoted, rustc_query_impl[d90372ddb6f5ed6d]::plumbing::QueryCtxt>
  32:     0x7ffaf4183e24 - <rustc_query_impl[d90372ddb6f5ed6d]::Queries as rustc_middle[114b3b537d7176f2]::ty::query::QueryEngine>::mir_promoted
  33:     0x7ffaf3ae2618 - rustc_borrowck[d9d24144480d32de]::mir_borrowck
  34:     0x7ffaf3aad840 - <rustc_borrowck[d9d24144480d32de]::provide::{closure#0} as core[227267dacc15a5d0]::ops::function::FnOnce<(rustc_middle[114b3b537d7176f2]::ty::context::TyCtxt, rustc_span[e0114d164763424]::def_id::LocalDefId)>>::call_once
  35:     0x7ffaf4519d62 - rustc_query_system[2b8658e7d9d902a0]::query::plumbing::try_execute_query::<rustc_query_impl[d90372ddb6f5ed6d]::plumbing::QueryCtxt, rustc_query_system[2b8658e7d9d902a0]::query::caches::VecCache<rustc_span[e0114d164763424]::def_id::LocalDefId, &rustc_middle[114b3b537d7176f2]::mir::query::BorrowCheckResult>>
  36:     0x7ffaf45cabd9 - rustc_query_system[2b8658e7d9d902a0]::query::plumbing::get_query::<rustc_query_impl[d90372ddb6f5ed6d]::queries::mir_borrowck, rustc_query_impl[d90372ddb6f5ed6d]::plumbing::QueryCtxt>
  37:     0x7ffaf41a3210 - <rustc_query_impl[d90372ddb6f5ed6d]::Queries as rustc_middle[114b3b537d7176f2]::ty::query::QueryEngine>::mir_borrowck
  38:     0x7ffaf28a1e41 - <core[227267dacc15a5d0]::panic::unwind_safe::AssertUnwindSafe<rustc_data_structures[89239e8ccb8fd3af]::sync::par_for_each_in<&[rustc_span[e0114d164763424]::def_id::LocalDefId], <rustc_middle[114b3b537d7176f2]::hir::map::Map>::par_body_owners<rustc_interface[97a011ea2c57c08b]::passes::analysis::{closure#2}::{closure#0}>::{closure#0}>::{closure#0}::{closure#0}> as core[227267dacc15a5d0]::ops::function::FnOnce<()>>::call_once
  39:     0x7ffaf27f6d26 - std[4d8c08a1d6cd93f8]::panic::catch_unwind::<core[227267dacc15a5d0]::panic::unwind_safe::AssertUnwindSafe<rustc_data_structures[89239e8ccb8fd3af]::sync::par_for_each_in<&[rustc_span[e0114d164763424]::def_id::LocalDefId], <rustc_middle[114b3b537d7176f2]::hir::map::Map>::par_body_owners<rustc_interface[97a011ea2c57c08b]::passes::analysis::{closure#2}::{closure#0}>::{closure#0}>::{closure#0}::{closure#0}>, ()>
  40:     0x7ffaf27edacd - rustc_data_structures[89239e8ccb8fd3af]::sync::par_for_each_in::<&[rustc_span[e0114d164763424]::def_id::LocalDefId], <rustc_middle[114b3b537d7176f2]::hir::map::Map>::par_body_owners<rustc_interface[97a011ea2c57c08b]::passes::analysis::{closure#2}::{closure#0}>::{closure#0}>
  41:     0x7ffaf2803235 - <rustc_session[c234fdae5de6c3ad]::session::Session>::time::<(), rustc_interface[97a011ea2c57c08b]::passes::analysis::{closure#2}>
  42:     0x7ffaf281b51b - rustc_interface[97a011ea2c57c08b]::passes::analysis
  43:     0x7ffaf44e3e20 - rustc_query_system[2b8658e7d9d902a0]::query::plumbing::try_execute_query::<rustc_query_impl[d90372ddb6f5ed6d]::plumbing::QueryCtxt, rustc_query_system[2b8658e7d9d902a0]::query::caches::DefaultCache<(), core[227267dacc15a5d0]::result::Result<(), rustc_errors[7a3c0cce756dd281]::ErrorGuaranteed>>>
  44:     0x7ffaf4633e04 - rustc_query_system[2b8658e7d9d902a0]::query::plumbing::get_query::<rustc_query_impl[d90372ddb6f5ed6d]::queries::analysis, rustc_query_impl[d90372ddb6f5ed6d]::plumbing::QueryCtxt>
  45:     0x7ffaf41774ca - <rustc_query_impl[d90372ddb6f5ed6d]::Queries as rustc_middle[114b3b537d7176f2]::ty::query::QueryEngine>::analysis
  46:     0x7ffaf272b6bc - <rustc_interface[97a011ea2c57c08b]::passes::QueryContext>::enter::<rustc_driver[5df8529dbce8ebc5]::run_compiler::{closure#1}::{closure#2}::{closure#2}, core[227267dacc15a5d0]::result::Result<(), rustc_errors[7a3c0cce756dd281]::ErrorGuaranteed>>
  47:     0x7ffaf2735aaf - <rustc_interface[97a011ea2c57c08b]::interface::Compiler>::enter::<rustc_driver[5df8529dbce8ebc5]::run_compiler::{closure#1}::{closure#2}, core[227267dacc15a5d0]::result::Result<core[227267dacc15a5d0]::option::Option<rustc_interface[97a011ea2c57c08b]::queries::Linker>, rustc_errors[7a3c0cce756dd281]::ErrorGuaranteed>>
  48:     0x7ffaf26d1c26 - rustc_span[e0114d164763424]::with_source_map::<core[227267dacc15a5d0]::result::Result<(), rustc_errors[7a3c0cce756dd281]::ErrorGuaranteed>, rustc_interface[97a011ea2c57c08b]::interface::run_compiler<core[227267dacc15a5d0]::result::Result<(), rustc_errors[7a3c0cce756dd281]::ErrorGuaranteed>, rustc_driver[5df8529dbce8ebc5]::run_compiler::{closure#1}>::{closure#0}::{closure#0}>
  49:     0x7ffaf273683a - <scoped_tls[849b60d84d59a484]::ScopedKey<rustc_span[e0114d164763424]::SessionGlobals>>::set::<rustc_interface[97a011ea2c57c08b]::interface::run_compiler<core[227267dacc15a5d0]::result::Result<(), rustc_errors[7a3c0cce756dd281]::ErrorGuaranteed>, rustc_driver[5df8529dbce8ebc5]::run_compiler::{closure#1}>::{closure#0}, core[227267dacc15a5d0]::result::Result<(), rustc_errors[7a3c0cce756dd281]::ErrorGuaranteed>>
  50:     0x7ffaf26f009f - std[4d8c08a1d6cd93f8]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[97a011ea2c57c08b]::util::run_in_thread_pool_with_globals<rustc_interface[97a011ea2c57c08b]::interface::run_compiler<core[227267dacc15a5d0]::result::Result<(), rustc_errors[7a3c0cce756dd281]::ErrorGuaranteed>, rustc_driver[5df8529dbce8ebc5]::run_compiler::{closure#1}>::{closure#0}, core[227267dacc15a5d0]::result::Result<(), rustc_errors[7a3c0cce756dd281]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[227267dacc15a5d0]::result::Result<(), rustc_errors[7a3c0cce756dd281]::ErrorGuaranteed>>
  51:     0x7ffaf274fb56 - std[4d8c08a1d6cd93f8]::panicking::try::<core[227267dacc15a5d0]::result::Result<(), rustc_errors[7a3c0cce756dd281]::ErrorGuaranteed>, core[227267dacc15a5d0]::panic::unwind_safe::AssertUnwindSafe<<std[4d8c08a1d6cd93f8]::thread::Builder>::spawn_unchecked_<rustc_interface[97a011ea2c57c08b]::util::run_in_thread_pool_with_globals<rustc_interface[97a011ea2c57c08b]::interface::run_compiler<core[227267dacc15a5d0]::result::Result<(), rustc_errors[7a3c0cce756dd281]::ErrorGuaranteed>, rustc_driver[5df8529dbce8ebc5]::run_compiler::{closure#1}>::{closure#0}, core[227267dacc15a5d0]::result::Result<(), rustc_errors[7a3c0cce756dd281]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[227267dacc15a5d0]::result::Result<(), rustc_errors[7a3c0cce756dd281]::ErrorGuaranteed>>::{closure#1}::{closure#0}>>
  52:     0x7ffaf26e09d5 - <<std[4d8c08a1d6cd93f8]::thread::Builder>::spawn_unchecked_<rustc_interface[97a011ea2c57c08b]::util::run_in_thread_pool_with_globals<rustc_interface[97a011ea2c57c08b]::interface::run_compiler<core[227267dacc15a5d0]::result::Result<(), rustc_errors[7a3c0cce756dd281]::ErrorGuaranteed>, rustc_driver[5df8529dbce8ebc5]::run_compiler::{closure#1}>::{closure#0}, core[227267dacc15a5d0]::result::Result<(), rustc_errors[7a3c0cce756dd281]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[227267dacc15a5d0]::result::Result<(), rustc_errors[7a3c0cce756dd281]::ErrorGuaranteed>>::{closure#1} as core[227267dacc15a5d0]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  53:     0x7ffaf1c97dce - std::sys::unix::thread::Thread::new::thread_start::hfb46186e3691734c
  54:     0x7ffaf1a2db43 - <unknown>
  55:     0x7ffaf1abfa00 - <unknown>
  56:                0x0 - <unknown>
note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md


note: rustc 1.67.0-nightly (b19a9586f 2022-12-01) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z threads=1 -C opt-level=1 -Z dump-mir=all -Z validate-mir -Z dump-mir-exclude-pass-number -Z mir-pretty-relative-line-numbers=yes -Z mir-opt-level=4 -Z dump-mir-dir=/checkout/obj/build/x86_64-unknown-linux-gnu/test/mir-opt/building/custom/consts -C prefer-dynamic -C rpath -C debuginfo=0
query stack during panic:
query stack during panic:
#0 [unsafety_check_result] unsafety-checking `consts`
#1 [mir_const] preparing `consts` for borrow checking
#2 [mir_promoted] processing MIR for `consts`
#3 [mir_borrowck] borrow-checking `consts`
#4 [analysis] running analysis passes on this crate
error: aborting due to previous error
------------------------------------------


```

</details>

nbdd0121 committed:
Ensure valid local_data is set for custom mir building

MIR unsafety checking requires this to be valid

nbdd0121 committed:
Remove unnecessary recursive call to parent unsafeck

All bodies are unsafe checked anyway. Current MIR unsafeck also just
returns for closures.

nbdd0121 committed:
Add tests (currently broken) for unsafe + inline const pat

RalfJung commented:
> Note that this PR does not fix the issue for unsafe { [0; function_requiring_unsafe()] }. This is fundamentally unfixable for MIR unsafeck IMO.

Why is that any harder than having `unsafe` propagate into closures? Both are completely separate MIR bodies after all. So somehow the MIR body for the closure is informed that is "lives" in an `unsafe` block, and the same could happen with array lengths?

RalfJung commented on compiler/rustc_mir_build/src/check_unsafety.rs:
```diff
@@ -408,16 +408,29 @@ impl<'a, 'tcx> Visitor<'a, 'tcx> for UnsafetyVisitor<'a, 'tcx> {
+                    let mut inner_visitor =
+                        UnsafetyVisitor { thir: inner_thir, hir_context, ..*self };
+                    inner_visitor.visit_expr(&inner_thir[expr]);
+                    // Unsafe blocks can be used in closures, make sure to take it into account
```
```suggestion
                    // Unsafe blocks can be used in inline consts, make sure to take it into account
```

nbdd0121 commented:
> Why is that any harder than having `unsafe` propagate into closures? Both are completely separate MIR bodies after all. So somehow the MIR body for the closure is informed that is "lives" in an `unsafe` block, and the same could happen with array lengths?

The difficulty arises precisely from the fact that they are separate MIR bodies. For closures (and after this PR, inline consts), their use is visible in the MIR of the containing body -- closures appear as a `Rvalue::Aggregate` and inline consts appear as `Operand::Constant`. However the same thing cannot be said for array length. Because array length feeds into the type system, it was (and needs to be) simplified and evaluated before the MIR is built, and the MIR built will reference these constants anymore.

So the issue for anon consts cannot easily be resolved with MIR unsafeck. It should be solvable, though, once we move to THIR unsafeck.

oli-obk commented on compiler/rustc_mir_build/src/check_unsafety.rs:
```diff
@@ -408,16 +408,29 @@ impl<'a, 'tcx> Visitor<'a, 'tcx> for UnsafetyVisitor<'a, 'tcx> {
-                closure_visitor.visit_expr(&closure_thir[expr]);
-                // Unsafe blocks can be used in closures, make sure to take it into account
-                self.safety_context = closure_visitor.safety_context;
+                if let Ok((closure_thir, expr)) = self.tcx.thir_body(closure_def) {
+                    let closure_thir = &closure_thir.borrow();
+                    let hir_context = self.tcx.hir().local_def_id_to_hir_id(closure_id);
+                    let mut closure_visitor =
+                        UnsafetyVisitor { thir: closure_thir, hir_context, ..*self };
+                    closure_visitor.visit_expr(&closure_thir[expr]);
+                    // Unsafe blocks can be used in closures, make sure to take it into account
+                    self.safety_context = closure_visitor.safety_context;
+                }
```
could this be a method called for closures and const blocks instead of having (almost?) the same code twice?

nbdd0121 replied:
Done

nagisa commented:
r? @oli-obk 

I can’t even begin to guess whether we actually want to do this…

nbdd0121 committed:
Extract shared logic into a new function

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit d6dc9124b7b814c2c5c35aa6295ea5e54071daac with merge 26e606eb2c7253f83a3db16651f9ed5443c7d61d...
<!-- homu: {"type":"TryBuildStarted","head_sha":"d6dc9124b7b814c2c5c35aa6295ea5e54071daac","merge_sha":"26e606eb2c7253f83a3db16651f9ed5443c7d61d"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3684058954/jobs/6235486289)
Build commit: 26e606eb2c7253f83a3db16651f9ed5443c7d61d (`26e606eb2c7253f83a3db16651f9ed5443c7d61d`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3684058954/jobs/6235486289"},"merge_sha":"26e606eb2c7253f83a3db16651f9ed5443c7d61d"} -->

rust-timer commented:
Finished benchmarking commit (26e606eb2c7253f83a3db16651f9ed5443c7d61d): [comparison URL](https://perf.rust-lang.org/compare.html?start=71ec1457ee9868a838e4521a3510cdd416c0c295&end=26e606eb2c7253f83a3db16651f9ed5443c7d61d&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=71ec1457ee9868a838e4521a3510cdd416c0c295&end=26e606eb2c7253f83a3db16651f9ed5443c7d61d&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.1% | [1.0%, 1.2%] | 2     |
| Regressions ❌ <br /> (secondary)  | 2.3% | [2.1%, 2.7%] | 6     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.1% | [1.0%, 1.2%] | 2     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=71ec1457ee9868a838e4521a3510cdd416c0c295&end=26e606eb2c7253f83a3db16651f9ed5443c7d61d&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.4%  | [2.9%, 3.8%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 2.3%  | [2.3%, 2.3%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.0% | [-2.0%, -2.0%] | 1     |
| All ❌✅ (primary)                 | 3.4%  | [2.9%, 3.8%]   | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=71ec1457ee9868a838e4521a3510cdd416c0c295&end=26e606eb2c7253f83a3db16651f9ed5443c7d61d&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

oli-obk commented:
These regressions are just noise, the exact same regression is happening on unrelated other PRs.

@bors r+ rollup
